File: src/test/java/com/polarbookshop/orderservice/order/event/OrderFunctionsTest.java
package com.polarbookshop.orderservice.order.event;

import static org.assertj.core.api.Assertions.assertThat;

import com.polarbookshop.orderservice.TestcontainersConfiguration;
import com.polarbookshop.orderservice.order.domain.Order;
import com.polarbookshop.orderservice.order.domain.OrderRepository;
import com.polarbookshop.orderservice.order.domain.OrderStatus;
import java.time.Duration;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.cloud.stream.binder.test.InputDestination;
import org.springframework.cloud.stream.binder.test.TestChannelBinderConfiguration;
import org.springframework.context.annotation.Import;
import org.springframework.integration.support.MessageBuilder;
import org.testcontainers.junit.jupiter.Testcontainers;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

@SpringBootTest
@Import({TestChannelBinderConfiguration.class, TestcontainersConfiguration.class})
@Testcontainers
class OrderFunctionsTest {

  @Autowired InputDestination input;

  @Autowired OrderRepository orderRepository;

  @BeforeEach
  void setUp() {
    orderRepository.deleteAll().block();
  }

  @Test
  void dispatchOrder() {
    Order order =
        orderRepository
            .save(Order.create("12345678890", "Book", 9.90, 3, OrderStatus.ACCEPTED))
            .block();

    var inputMessage = MessageBuilder.withPayload(new OrderDispatchedMessage(order.id())).build();
    input.send(inputMessage);

    Mono<Order> eventuallyDispatched =
        orderRepository
            .findById(order.id())
            .filter(o -> o.status().equals(OrderStatus.DISPATCHED))
            .switchIfEmpty(Mono.empty())
            .repeatWhenEmpty(repeat -> repeat.delayElements(Duration.ofMillis(50)))
            .timeout(Duration.ofSeconds(3));

    eventuallyDispatched
        .as(StepVerifier::create)
        .assertNext(
            dispatched -> {
              assertThat(dispatched.id()).isEqualTo(order.id());
              assertThat(dispatched.status()).isEqualTo(OrderStatus.DISPATCHED);
            })
        .verifyComplete();
  }

  @Test
  void dispatchOrder_orderNotFound() {
    var inputMessage = MessageBuilder.withPayload(new OrderDispatchedMessage(999L)).build();
    input.send(inputMessage);

    Mono<Long> eventuallyCount = Mono.delay(Duration.ofMillis(300)).then(orderRepository.count());

    eventuallyCount.as(StepVerifier::create).expectNext(0L).verifyComplete();
  }
}

File: src/test/kotlin/com/polarbookshop/orderservice/order/event/OrderFunctionsTest.kt
package com.polarbookshop.orderservice.order.event

import com.polarbookshop.orderservice.order.domain.Order
import com.polarbookshop.orderservice.order.domain.OrderService
import com.polarbookshop.orderservice.order.domain.OrderStatus
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import reactor.core.publisher.Flux
import java.time.Instant
import java.util.concurrent.atomic.AtomicBoolean

@ExtendWith(MockKExtension::class)
class OrderFunctionsTest {
    @MockK
    lateinit var orderService: OrderService

    @Test
    fun `dispatch order`() {
        val orderId = 121L
        val messages = Flux.just(OrderDispatchedMessage(orderId))
        val subscribed = AtomicBoolean()
        val orders =
            Flux
                .just(
                    Order(
                        bookIsbn = "1234567890",
                        bookName = "The Hobbit",
                        bookPrice = 9.90,
                        quantity = 1,
                        status = OrderStatus.DISPATCHED,
                        id = orderId,
                        createdDate = Instant.now(),
                        lastModifiedDate = Instant.now(),
                        version = 21,
                    ),
                ).doOnSubscribe { subscribed.set(true) }
        every { orderService.consumeOrderDispatchedEvent(any()) } returns orders

        OrderFunctions().dispatchOrder(orderService).accept(messages)

        verify(exactly = 1) { orderService.consumeOrderDispatchedEvent(any()) }
        assertThat(subscribed).isTrue
    }
}
